### PR TITLE
Event prediction

### PR DESCRIPTION
**Pull request by jorshi**
For event predictions this converts a set of frame based label predictions to a list of events with start and stop timestamps in ms.

### CODE REVIEW

File: heareval/predictions/task_predictions.py [hunk 1]
```diff
@@ -30,8 +32,9 @@ def __init__(self, nfeatures: int, nlabels: int, prediction_type: str):
         super().__init__()
 
         self.projection = torch.nn.Linear(nfeatures, nlabels)
+        torch.nn.init.normal_(self.projection.weight)
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:35]: Was experimenting with different inits for testing -- can remove this.

File: heareval/predictions/task_predictions.py [hunk 2]
```diff
@@ -120,9 +234,34 @@ def task_predictions(
             predicted_labels = predictor(embs)
             # TODO: Uses less memory to stack them one at a time
             all_predicted_labels.append(predicted_labels)
-        all_predicted_labels = torch.cat(all_predicted_labels)
+        predicted_labels = torch.cat(all_predicted_labels)
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:237]: mypy

File: heareval/predictions/task_predictions.py [hunk 3]
```diff
@@ -79,6 +82,117 @@ def __getitem__(self, idx):
...
+            # New onset
+            if pred_diff[t][class_idx] == 1:
+                assert len(current_event) == 0
+                current_event = [timestamps[t]]
+
+            # Offset for current event
+            elif pred_diff[t][class_idx] == -1:
+                assert len(current_event) == 1
+                # TODO: filter on event length? DCASE only included events that
+                #   are 60ms long.
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:133]: **atleast** 60ms

File: heareval/predictions/task_predictions.py (hunk 1 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:35]: okay, I'm fine either way

File: heareval/predictions/task_predictions.py [hunk 4]
```diff
@@ -79,6 +82,117 @@ def __getitem__(self, idx):
         # return self.embedding_memmap[idx], self.labels[idx]
```
Comment (turian) [inline heareval/predictions/task_predictions.py:84]: On the one hand, I'm not sure we want to do this here. On the other hand, I have no idea where we want to do this? So I guess this is fine :)

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:133]: This is worth a discussion :\

File: heareval/predictions/task_predictions.py [hunk 5]
```diff
@@ -79,6 +82,117 @@ def __getitem__(self, idx):
...
+    # TODO: Apply median filtering for smoothing?
+    #   Additionally, DCASE limits the number of active classes at a single
+    #   time to 6. We could do that here as well.
+    predictions = (predictions > threshold).type(torch.int8)
+
+    # Difference between each timestamp to find event boundaries
+    pred_diff = torch.diff(predictions, dim=0)
+
+    # This is slow, but for each class look through all the timestamp predictions
+    # and construct a set of events with onset and offset timestamps.
```
Comment (turian) [inline heareval/predictions/task_predictions.py:114]: I really am not sure what this is doing, is it what `more_itertools.consecutive_groups` does? https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.consecutive_groups

File: heareval/predictions/task_predictions.py [hunk 6]
```diff
@@ -79,6 +82,117 @@ def __getitem__(self, idx):
         # return self.embedding_memmap[idx], self.labels[idx]
 
 
+def create_events_from_prediction(
+    prediction_dict: Dict, threshold: float = 0.5
```
Comment (turian) [inline heareval/predictions/task_predictions.py:86]: Maybe since we expect the key to be float timestamp, etc. put that in the docstring and type here?

Comment (turian) [reply heareval/predictions/task_predictions.py:86]: Like Dict[float, float]

Comment (turian) [reply heareval/predictions/task_predictions.py:86]: I still am not 100% sure what the signature is for prediction dict. Does it involve the label or index? Is there are probability?

File: heareval/predictions/task_predictions.py (hunk 5 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:114]: It's based on the dcase code - it's basically looking for boundaries between consecutive groups of the *same* element. It achieves that by using the difference function above. So you get a value of 1 when you have an onset and then -1 for an onset because we have converted to binary vectors. So it starts at the first timestamp and loops through looking for the boundaries and adds an event when there has been an onset and then an offset.

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:133]: Ya -- could be defined for each task

File: heareval/predictions/task_predictions.py (hunk 6 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py]: Updated this : Dict[float, torch.Tensor]

File: heareval/predictions/task_predictions.py [hunk 7]
```diff
@@ -79,6 +83,133 @@ def __getitem__(self, idx):
...
 
 
+def create_events_from_prediction(
+    prediction_dict: Dict[float, torch.Tensor],
+    threshold: float = 0.5,
+    median_filter_ms=90.0,
+    min_duration=60.0,
+) -> IntervalTree:
+    """
+    Takes a set of prediction tensors keyed on timestamps and generates events.
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:93]: Better decsription

File: heareval/predictions/task_predictions.py [hunk 8]
```diff
@@ -79,6 +83,133 @@ def __getitem__(self, idx):
...
+    timestamps = np.array(sorted(prediction_dict.keys()))
+
+    # Create a sorted numpy matrix of frame level predictions for this file. We convert
+    # to a numpy array here before applying a median filter.
+    predictions = np.stack([prediction_dict[t].detach().numpy() for t in timestamps])
+
+    # Apply median filtering and then thresholding to get the
+    # label detected at each frame.
+    # TODO: DCASE limit the number of concurrent events to 5 - should we?
+    ts_diff = timestamps[1] - timestamps[0]
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:117]: Check that all timestamps are the same

File: heareval/predictions/task_predictions.py [hunk 9]
```diff
@@ -79,6 +83,133 @@ def __getitem__(self, idx):
...
+    filter_width = int(round(median_filter_ms / ts_diff))
+
+    # Apply filter along the frames for each class to smooth out the labels
+    predictions = median_filter(predictions, size=(filter_width, 1))
+    predictions = (predictions > threshold).astype(np.int8)
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:122]: Add a comment -- binary vector

File: heareval/predictions/task_predictions.py [hunk 10]
```diff
@@ -79,6 +83,133 @@ def __getitem__(self, idx):
...
+    predictions = np.stack([prediction_dict[t].detach().numpy() for t in timestamps])
+
+    # Apply median filtering and then thresholding to get the
+    # label detected at each frame.
+    # TODO: DCASE limit the number of concurrent events to 5 - should we?
+    ts_diff = timestamps[1] - timestamps[0]
+    filter_width = int(round(median_filter_ms / ts_diff))
+
+    # Apply filter along the frames for each class to smooth out the labels
+    predictions = median_filter(predictions, size=(filter_width, 1))
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:121]: Remove filter -- can give an unfair advantage depending on timestamp interval

File: heareval/predictions/task_predictions.py [hunk 11]
```diff
@@ -79,6 +83,133 @@ def __getitem__(self, idx):
...
+                if event_duration >= min_duration:
+                    event_tree.addi(
+                        begin=current_event[0], end=current_event[1], data=class_idx
+                    )
+                current_event = []
+
+    return event_tree
+
+
+def get_predictions_as_events(
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:159]: Better names or more detailed comments

**Review by turian — APPROVED** [review]
